File: leetcode_75/HashMap_Set/python/2352.py
from typing import List


class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        n = len(grid)
        strings = dict()
        for line in grid:
            hashable_line = tuple(line)
            if hashable_line not in strings:
                strings[hashable_line] = 0
            strings[hashable_line] += 1

        counter = 0
        for i in range(n):
            cur_column = list()
            for j in range(n):
                cur_column.append(grid[j][i])
            hashable_column = tuple(cur_column)
            if hashable_column in strings:
                counter += strings[hashable_column]

        return counter

File: leetcode_75/HashMap_Set/cpp/2352.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution
{
public:
    int equalPairs(vector<vector<int>> &grid)
    {
        map<vector<int>, int> strings;
        for (auto line : grid)
        {
            if (strings.find(line) == strings.end())
            {
                strings[line] = 0;
            }
            strings[line]++;
        }

        int counter = 0;
        int n = grid.size();
        for (int i = 0; i < n; ++i)
        {
            vector<int> cur_column;
            for (int j = 0; j < n; ++j)
            {
                cur_column.push_back(grid[j][i]);
            }
            if (strings.find(cur_column) != strings.end())
            {
                counter += strings[cur_column];
            }
        }

        return counter;
    }
};